qemu/tests/watchdog.py

```diff
@@ -282,6 +283,52 @@ def hotplug_unplug_watchdog_device():
                 test.fail("Oops, find watchdog pci, unplug failed")
             logging.info("The WDT remove successfully")
 
+    def heartbeat_test():
+        """
+        Heartbeat test for i6300esb
+        Test steps:
+        1.Start VM with "-watchdog-action pause" CLI option
+        2.Set heartbeat value and reload the i6300esb module
+        3.Trigger wathchdog action through open /dev/watchdog
+        4.Ensure watchdog_action takes effect after $heartbeat.
+        """
+
+        del_module_cmd = params["del_module_cmd"]
+        reload_module_cmd = params["reload_module_cmd"]
+        _watchdog_device_check(test, session, watchdog_device_type)
+
+        error_context.context("set heartbeat value and reload the i6300esb "
+                              "module", logging.info)
+        session.cmd(del_module_cmd)
+        heartbeat = random.randint(1, 180)
+        session.cmd_output(reload_module_cmd % heartbeat)
+        _watchdog_device_check(test, session, watchdog_device_type)
+
+        _trigger_watchdog(session, trigger_cmd)
```

Comment:
Should we put this trigger below line 310 -- start_time?
If I understand this right, you're trying to measure whether the watchdog expire time fits the value you set, that is to say whether it equals 'heartbeat'.
If so, put _trigger_watchdog below start_time is more accurate.
Correct me if I'm wrong about your intention.

Reply:
Yes, your understand is right. But logically we should first trigger watchdog, them start to calculate time. And because the code runs so fast, there is no big difference whether put it below line 310 or above.